### PR TITLE
(#65) support user selectable list of ciphers

### PR DESCRIPTION
**Pull request by vjanelle**
Mitigation for https://github.com/choria-io/go-network-broker/issues/75

Allow runtime list of user selectable ciphers, curves, and default to sensible values

### CODE REVIEW

File: filesec/ciphersuite.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+	tls.TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA:      "TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA",
+	tls.TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256:   "TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256",
+	tls.TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256: "TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256",
+	tls.TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA:      "TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA",
+	tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256:   "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256",
+	tls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256: "TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256",
+	tls.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384:   "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384",
+	tls.TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384: "TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384",
+	tls.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305:    "TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305",
+	tls.TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305:  "TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305",
```
Comment (ripienaar) [inline filesec/ciphersuite.go]: filter this down to just what your security people will be happy with, I trust their judgement and its easy for you

File: filesec/ciphersuite.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+func defaultCipherSuites() []uint16 {
+	return []uint16{
+		tls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
+		tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
+		tls.TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,
+		tls.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,
+		tls.TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,
+		tls.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
+	}
+}
```
Comment (ripienaar) [inline filesec/ciphersuite.go]: we can default this to what your security people want, we probably trust their judgement

File: filesec/ciphersuite.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+	"TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA":      tls.TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,
+	"TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256":   tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
+	"TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256": tls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
+	"TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384":   tls.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
+	"TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384": tls.TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,
+	"TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305":    tls.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,
+	"TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305":  tls.TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,
+}
+
+var cipherMapByID = map[uint16]string{
```
Comment (ripienaar) [inline filesec/ciphersuite.go]: This is not used I think? If not kill it.

File: filesec/ciphersuite.go (hunk 2 above)
Comment (vjanelle) [reply filesec/ciphersuite.go]: I'm ok with chacha being enabled, especially for speed tests.

File: filesec/ciphersuite.go (hunk 3 above)
Comment (vjanelle) [reply filesec/ciphersuite.go]: Is now - using the information for debug logging.

File: filesec/ciphersuite.go (hunk 2 above)
Comment (vjanelle) [reply filesec/ciphersuite.go]: It's also lower on the priority list I think, so a client would have to specifically opt into it.

File: filesec/ciphersuite.go (hunk 3 above)
Comment (ripienaar) [reply filesec/ciphersuite.go]: pity, maintaining a double list is meh :) can we generate it from the other one? Or write a helper to walk the other one by value?

Comment (ripienaar) [reply filesec/ciphersuite.go]: Like, its a mistake waiting to happen - they'll fall out of sync, so generating it or searching it - while a bit slower - is safer.

**Comment by ripienaar** [comment]
grab latest go-config commit hash into go mod here and lets see how it goes

**Pull request merged**